content/6_Advanced/Parallel_Binary_Search.mdx

```diff
@@ -0,0 +1,140 @@
+---
+id: parallel-binary-search
+title: 'Parallel Binary Search'
+author: Muaath Alqarni
+description: 'Answer binary search queries offline'
+prerequisites:
+  - binary-search
+  - divide-and-conquer
+frequency: 0
+---
+
+
+## Resources
+
+<Resources>
+	<Resource source="CF" title="Parallel Binary Seach [tutorial]" url="45578" />
+	<Resource source="SPBook" title="Parallel Binary Seach" url="https://kostka.dev/sp/spbook.pdf#page=22" />
+</Resources>
+
+<FocusProblem problem="sam" />
+
+## Solution - New Roads Queries
+We can see that connectivity of the nodes is a monotonic function.
+
+Now let's use parallel binary search:
+
+First, sort the queries by median.
+
+Then sweep through the array of edges and union them in the DSU.
+Whenever there is a query with median = $i$, if the two nodes where connected in DSU, it means the answer is $\leq i$, otherwise it means $\gt i$.
+
+**Complexity:**
+
+We need to do the sweep $O(\log_2{N})$ times, same as binary search.
+Sorting the queries by median is $O(Q \cdot \log_2{Q})$
+And $O(\alpha(N))$ for the DSU
+
+
+Total Complexity: $O((N + Q) \cdot \log_2(N+Q)^2 \cdot \alpha(N))$
+
+## Implementation - New Roads Queries
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int N = 2e5 + 1;
+
+struct DSU {
+	int cnt[N], par[N];
+	void Init(int sz) {
+		for (int i = 0; i <= sz; i++) par[i] = i, cnt[i] = 1;
+	}
+	int Root(int node) {
+		if (par[node] == node) return node;
+		return Root(par[node]);
+	}
+	void Union(int x, int y) {
+		x = Root(x), y = Root(y);
+		if (cnt[x] < cnt[y]) swap(x, y);
+		if (x == y) return;
+		par[y] = x;
+		cnt[x] += cnt[y];
+	}
+	bool Connected(int x, int y) { return Root(x) == Root(y); }
+};
+
+// Parallel Binary Search Query
+struct PBS {
+	// l, r are the binary search variables
+	int idx, l, r;
+	int a, b;
```

Comment:
rename to `u` and `v`, just standard convention

Reply:
i mean depends on the problem statement
unless this is an arbitrary struct

Reply:
I used a,b same as the problem statement